src/bp/nlu-core/entities/duckling-extractor/index.ts

```diff
@@ -0,0 +1,190 @@
+import { NLU } from 'botpress/sdk'
+import { ensureFile, pathExists, readJSON, writeJson } from 'fs-extra'
+import _ from 'lodash'
+import lru from 'lru-cache'
+import ms from 'ms'
+import sizeof from 'object-sizeof'
+import path from 'path'
+
+import { extractPattern } from '../../tools/patterns-utils'
+import { SPACE } from '../../tools/token-utils'
+import { EntityExtractionResult, SystemEntityExtractor } from '../../typings'
+
+import { DucklingParams, DucklingProvider } from './duckling-provider'
+import { mapDucklingToEntity } from './map-duckling'
+import { DucklingDimension } from './typings'
+
+interface KeyedItem {
+  input: string
+  idx: number
+  entities?: EntityExtractionResult[]
+}
+
+export const JOIN_CHAR = `::${SPACE}::`
+const BATCH_SIZE = 10
+
+export const DUCKLING_ENTITIES: DucklingDimension[] = [
+  'amountOfMoney',
+  'distance',
+  'duration',
+  'email',
+  'number',
+  'ordinal',
+  'phoneNumber',
+  'quantity',
+  'temperature',
+  'time',
+  'url',
+  'volume'
+]
+
+const CACHE_PATH = path.join(process.APP_DATA_PATH || '', 'cache', 'sys_entities.json')
+
+// Further improvements:
+// 1- in _extractBatch, shift results ==> don't walk whole array n times (nlog(n) vs n2)
+
+export class DucklingEntityExtractor implements SystemEntityExtractor {
+  public static enabled: boolean
+  private _provider: DucklingProvider
+
+  private static _cache: lru<string, EntityExtractionResult[]>
+  private _cacheDumpEnabled = true
+
+  constructor(private readonly logger?: NLU.Logger) {
+    this._provider = new DucklingProvider(logger)
+  }
+
+  public static get entityTypes(): string[] {
+    return DucklingEntityExtractor.enabled ? DUCKLING_ENTITIES : []
+  }
+
+  public static async configure(enabled: boolean, url: string, logger?: NLU.Logger) {
+    if (enabled) {
+      this.enabled = await DucklingProvider.init(url, logger)
```

Comment:
this has changed (extracted initialisation of duckling inside duckling provider)